src/btree/bt_vrfy.c

```diff
@@ -215,38 +215,55 @@ __wt_verify(WT_SESSION_IMPL *session, const char *cfg[])
 		WT_ERR(bm->checkpoint_load(bm, session,
 		    ckpt->raw.data, ckpt->raw.size,
 		    root_addr, &root_addr_size, true));
+		bm_unload = true;
+
+		/* Skip trees with no root page. */
+		if (root_addr_size != 0) {
+			/*
+			 * We have an exclusive lock on the handle, but we're
+			 * swapping root pages in-and-out of that handle, and
+			 * there's a race with eviction entering the tree and
+			 * seeing an invalid root page. Eviction must work on
+			 * trees being verified (else we'd have to do our own
+			 * eviction), lock eviction out while doing the setup.
+			 */
+			WT_ERR(__wt_evict_file_exclusive_on(session));
+			ret = __wt_btree_tree_open(
+			    session, root_addr, root_addr_size);
+			__wt_evict_file_exclusive_off(session);
+			WT_ERR(ret);
 
-		/*
-		 * Ignore trees with no root page.
-		 * Verify, then discard the checkpoint from the cache.
-		 */
-		if (root_addr_size != 0 &&
-		    (ret = __wt_btree_tree_open(
-		    session, root_addr, root_addr_size)) == 0) {
 			if (WT_VRFY_DUMP(vs))
 				WT_ERR(__wt_msg(session, "Root: %s %s",
 				    __wt_addr_string(session,
 				    root_addr, root_addr_size, vs->tmp1),
 				    __wt_page_type_string(
 				    btree->root.page->type)));
 
+			/* Verify the tree. */
 			WT_WITH_PAGE_INDEX(session,
 			    ret = __verify_tree(session, &btree->root, vs));
+			WT_ERR(ret);
 
-			WT_TRET(__wt_cache_op(session, WT_SYNC_DISCARD));
+			/* Discard the checkpoint from the cache. */
+			WT_ERR(__wt_cache_op(session, WT_SYNC_DISCARD));
 		}
 
 		/* Unload the checkpoint. */
-		WT_TRET(bm->checkpoint_unload(bm, session));
-		WT_ERR(ret);
+		bm_unload = false;
+		WT_ERR(bm->checkpoint_unload(bm, session));
```

Comment:
Do we need to do the ``__wt_evict_file_exclusive_on`` call before unloading the checkpoint? Eviction walks the ``dhqh`` list to find handles, so this handle is visitable at any time. Or is this only a problem if the eviction server sees a handle that is partially populated via ``__wt_btree_tree_open`` - if it comes in between now, and where you added the ``exclusive_on``, it'll see a NULL root page and do the right thing?